background-processor/src/lib.rs

```diff
@@ -0,0 +1,345 @@
+	/// resulting in the thread being terminated.
+	/// Important note: this thread will panic if invoking `persist_manager` results in an error (and
+	/// `start()` will need to be called again to restart the `BackgroundProcessor`).
+	/// There are 3 main options for handling this panic:
+	/// * wait on [`thread_handle`]'s `join()`, handle the error
+	/// * [configure] to abort on panic
+	/// * write a custom `persist_manager` to handle the error so it never gets returned to
+	///   `BackgroundProcessor`.
+	///
+	/// `persist_manager` is responsible for writing out the `ChannelManager` to disk, and/or uploading
+	/// to one or more backup services. See [`ChannelManager::write`] for writing out a `ChannelManager`.
+	/// See [`FilesystemPersister::persist_manager`] for Rust-Lightning's provided implementation.
+	///
+	/// [`thread_handle`]: struct.BackgroundProcessor.html#structfield.thread_handle
+	/// [configure]: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html
+	/// [`ChannelManager::write`]: ../lightning/ln/channelmanager/struct.ChannelManager.html#method.write
+	/// [`FilesystemPersister::persist_manager`]: ../lightning_persister/struct.FilesystemPersister.html#impl
+	#[must_use]
+	pub fn start<PM, ChanSigner, M, T, K, F, L>(persist_manager: PM, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>, logger: Arc<L>) -> Self
+	where ChanSigner: 'static + ChannelKeys + Writeable,
+        M: 'static + chain::Watch<Keys=ChanSigner>,
+        T: 'static + BroadcasterInterface,
+        K: 'static + KeysInterface<ChanKeySigner=ChanSigner>,
+        F: 'static + FeeEstimator,
+        L: 'static + Logger,
+        PM: 'static + Send + Fn(Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>,
+	{
+		let stop_thread = Arc::new(AtomicBool::new(false));
+		let stop_thread_clone = stop_thread.clone();
+		let handle = thread::spawn(move || {
+			let mut current_time = Instant::now();
+			loop {
+				let updates_available = manager.wait_timeout(Duration::from_millis(100));
+				if updates_available {
+					if let Err(e) = persist_manager(manager.clone()) {
+						panic!("Errored persisting manager: {}", e);
+					};
+				}
+				// If we see that the thread has been stopped, exit now.
+				if stop_thread.load(Ordering::Relaxed) == true {
+					log_trace!(logger, "Terminating background processor.");
+					break;
+				}
+				if current_time.elapsed().as_secs() > CHAN_FRESHNESS_TIMER {
+					log_trace!(logger, "Calling manager's timer_chan_freshness_every_min");
+					manager.timer_chan_freshness_every_min();
+					current_time = Instant::now();
+				}
+			}
+		});
+		Self {
+			stop_thread: stop_thread_clone,
+			thread_handle: handle
+		}
+	}
+
+	/// Stop `BackgroundProcessor`'s thread.
+	pub fn stop(&self) {
+		self.stop_thread.store(true, Ordering::Relaxed)
```

Comment:
Why not release-acquire on the stop_thread variable instead of relaxed?

Reply:
I guess it wasn't obvious to me that preserving the order of operations between threads was super crucial. I don't understand `atomic::Ordering` too well though. Is there a circumstance where it would matter here?

Reply:
Hmm, dunno, relaxed, I believe, is basically not guaranteed to run at all, and while rel-acq may be a bit slower, we're never using it in a performance-critical section, so it shouldn't matter.

Reply:
> relaxed, I believe, is basically not guaranteed to run at all

Ahh wasn't aware. Fixed I believe.